### DISCUSSION TITLE
conditional compilation routes and data.

### DISCUSSION THREAD

**Discussion by alkeryn** [actix/actix-web#2946] [Q&A] [answered]
Hey! is there a way that isn't too cumbersome to have some routes present only if compiled with a specific feature flag or stuff like that ?
we can't put a `#[cfg]` macro between two `.route()` so i wonder if there is a clean way to do it.
i thought of doing something like 
```rust
.configure(if cfg!(feature = "tests") { testroutes::get_routes } else { noop })
```
but that doesn't seem clean ^^
also i haven't found a good solution for app_data().

maybe it is possible to define App outside http server ?
or should we use a big configure function that has the conditions here ?

thank you ! :)

**Comment by robjtede** [comment, marked as answer]
```rust
use actix_web::{web, App, HttpRequest, HttpServer, Responder};

async fn greet(req: HttpRequest) -> impl Responder {
    let name = req.match_info().get("name").unwrap_or("World");
    format!("Hello {}!", &name)
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| {
        let app = App::new().route("/", web::get().to(greet));

        let app = if cfg!(feature = "on") {
            app.configure(|cfg| {
                cfg.route("/on", web::get().to(greet));
            })
        } else {
            app
        };

        let app = app.route("/{name}", web::get().to(greet));

        app
    })
    .bind(("127.0.0.1", 8080))?
    .run()
    .await
}

```

**Reply by alkeryn** [reply]
@robjtede but can we define app outside of HttpServer ?
if i try to the compiler just screams.

**Reply by alkeryn** [reply]
also doesn't really works, it complains about an use after move which is odd.
well redefining it seems to work.

**Reply by robjtede** [reply]
I've updated my original answer with something more correct and complete that compiles.

**Reply by robjtede** [reply]
Also, no you cannot create App outside the HttpServer closure since it is not `Clone` or `Send`.

**Reply by alkeryn** [reply]
hey thanks ! :)

**Reply by Cakeday** [reply]
just FYI, the cfg! macro doesn't conditionally compile: https://doc.rust-lang.org/std/macro.cfg.html.

**Reply by alkeryn** [reply]
@Cakeday whilst true, in the given example it unfolds to an if(true), but then the compiler optimisations will just remove the if part most likely because it is known at compile time that it will always be true so i feel like it still kind of is.
thanks though :)
